File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size(), leftProduct = nums[0], rightProduct = nums[n-1];
        vector<int> answer(n, 1);
        for(auto i=1; i<n; i++){
            answer[i] = leftProduct;
            leftProduct *= nums[i];
        }
        for(auto i=n-2; i>=0; i--){
            answer[i] *= rightProduct;
            rightProduct *= nums[i];
        }
        return answer;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] productArray = new int[nums.length];
        int leftProduct = 1, rightProduct = 1;
        Arrays.fill(productArray, 1);
        for(int i=0; i<nums.length; i++){
            productArray[i] = leftProduct;
            leftProduct *= nums[i];
        }
        for(int i=nums.length-1; i>=0; i--){
            productArray[i] *= rightProduct;
            rightProduct *= nums[i];
        }
        return productArray;
    }
}